text/0000-try-trait.md

```diff
@@ -0,0 +1,553 @@
+
+The original `Carrier` as proposed in RFC 243 had a rather different
+design:
+
+```rust
+trait ResultCarrier {
+    type Normal;
+    type Exception;
+    fn embed_normal(from: Normal) -> Self;
+    fn embed_exception(from: Exception) -> Self;
+    fn translate<Other: ResultCarrier<Normal=Normal, Exception=Exception>>(from: Self) -> Other;
+}
+```
+
+Whereas this `Try` trait links the type of the value being matched
+(`Self`) with the type that the enclosing function returns (`E`), the
+`Carrier` was implemented separately for each type (e.g., `Result` and
+`Option`). It allowed any kind of carrier to be converted to any other
+kind of carrier, which fails to preserve the "semantic equivalent"
+property.  It would also interact poorly with type inference, as
+discussed in the "Detailed Design" section.
+
+### Traits implemented over higher-kinded types
+
+The "semantic equivalent" property might suggest that the `Carrier`
+trait ought to be defined over higher-kinded types (or generic
+associated types) in some form. The most obvious downside of such a
+design is that Rust does not offer higher-kinded types nor anything
+equivalent to them today, and hence we would have to block on that
+design effort. But it also turns out that HKT is
+[not a particularly good fit for the problem](https://github.com/rust-lang/rust/pull/35056#issuecomment-240129923). To
+start, consider what "kind" the `Self` parameter on the `Try` trait
+would have to have.  If we were to implement `Try` on `Option`, it
+would presumably then have kind `type -> type`, but we also wish to
+implement `Try` on `Result`, which has kind `type -> type ->
+type`. There has even been talk of implementing `Try` for simple types
+like `bool`, which simply have kind `type`. More generally, the
+problems encountered are quite similar to the problems that
+[Simon Peyton-Jones describes in attempting to model collections using HKT](https://github.com/rust-lang/rust/pull/35056#issuecomment-240129923):
+we wish the `Try` trait to be implemented in a great number of
+scenarios.  Some of them, like converting `Result<T,E>` to
+`Result<U,F>`, allow for the type of the success value and the error
+value to both be changed, though not arbitrarily (subject to the
+`From` trait, in particular).  Others, like converting `Option<T>` to
+`Option<U>`, allow only the type of the success value to change,
+whereas others (like converting `bool` to `bool`) do not allow either
+type to change.
+
+### What to name the trait
+
+A number of names have been proposed for this trait. The original name
+was `Carrier`, as the implementing type was the "carrier" for an error
+value. A proposed alternative was `QuestionMark`, named after the
+operator `?`. However, the general consensus seemed to be that since
+Rust operator overloading traits tend to be named after the
+*operation* that the operator performed (e.g., `Add` and not `Plus`,
+`Deref` and not `Star` or `Asterix`), it was more appropriate to name
```

Comment:
It would be awesome to have Asterix as an operator, can we also have Obelix please? :D
However, I think you probably meant to write "Asterisk".